swarm/storage/resource.go

```diff
@@ -91,186 +114,168 @@ type resource struct {
+		rh.nameHash = func(name string) common.Hash {
 			rh.hashLock.Lock()
 			defer rh.hashLock.Unlock()
 			rh.hasher.Reset()
 			rh.hasher.Write([]byte(name))
 			return common.BytesToHash(rh.hasher.Sum(nil))
-		})
+		}
 	}
 
 	return rh, nil
 }
 
-func validateInput(name string, frequency uint64) (string, error) {
-	// frequency 0 is invalid
-	if frequency == 0 {
-		return "", fmt.Errorf("Frequency cannot be 0")
-	}
+func (self *ResourceHandler) IsValidated() bool {
+	return self.validator == nil
+}
 
-	// must have name
-	if name == "" {
-		return "", fmt.Errorf("Name cannot be empty")
-	}
+func (self *ResourceHandler) HashSize() int {
+	return self.validator.hashSize()
+}
 
-	// make sure our ens identifier is idna safe
-	validname, err := idna.ToASCII(name)
-	if err != nil {
-		return "", err
-	}
+// get data from current resource
 
-	return validname, nil
+func (self *ResourceHandler) GetContent(name string) (Key, []byte, error) {
+	rsrc := self.getResource(name)
+	if rsrc == nil || !rsrc.isSynced() {
+		return nil, nil, fmt.Errorf("Resource does not exist or is not synced")
+	}
+	return rsrc.lastKey, rsrc.data, nil
 }
 
-// Creates a standalone resource object
-//
-// Can be passed to SetResource if external root data lookups are used
-func NewResource(name string, startBlock uint64, frequency uint64, nameHashFunc func(name string) common.Hash) (*resource, error) {
+func (self *ResourceHandler) GetLastPeriod(name string) (uint32, error) {
+	rsrc := self.getResource(name)
 
-	validname, err := validateInput(name, frequency)
-	if err != nil {
-		return nil, err
+	if rsrc == nil || !rsrc.isSynced() {
+		return 0, fmt.Errorf("Resource does not exist or is not synced")
 	}
+	return rsrc.lastPeriod, nil
+}
 
-	return &resource{
-		name:       validname,
-		nameHash:   nameHashFunc(validname),
-		startBlock: startBlock,
-		frequency:  frequency,
-	}, nil
+func (self *ResourceHandler) GetVersion(name string) (uint32, error) {
+	rsrc := self.getResource(name)
+	if rsrc == nil || !rsrc.isSynced() {
+		return 0, fmt.Errorf("Resource does not exist or is not synced")
+	}
+	return rsrc.version, nil
+}
+
+// \TODO should be hashsize * branches from the chosen chunker, implement with dpa
+func (self *ResourceHandler) chunkSize() int64 {
+	return chunkSize
 }
 
 // Creates a new root entry for a mutable resource identified by `name` with the specified `frequency`.
 //
+// The signature data should match the hash of the idna-converted name by the validator's namehash function, NOT the raw name bytes.
+//
 // The start block of the resource update will be the actual current block height of the connected network.
 func (self *ResourceHandler) NewResource(name string, frequency uint64) (*resource, error) {
 
-	ok, err := self.validator.isOwner(name)
-	if err != nil {
-		return nil, err
-	} else if !ok {
-		return nil, fmt.Errorf("Not owner of '%s'", name)
+	// frequency 0 is invalid
+	if frequency == 0 {
+		return nil, fmt.Errorf("Frequency cannot be 0")
```

Comment:
Use `errors.New`.

Reply:
How was this policy again? I remember changing a bunch of these to `errors.New`, only having to change them back later...?

Reply:
`fmt.Errorf` is when you want to format an error (e.g. `fmt.Errorf("error: %s", err)`), but if you just want to use a string as an error, use `errors.New` to save trying to format nothing.

[Here](https://github.com/ethersphere/go-ethereum/pull/125#discussion_r143342413) is my previous comment.